### ISSUE TITLE
Add support for v1.2 engine/demo compatibility

### ISSUE THREAD

**Issue by fabiangreffrath** [chocolate-doom/chocolate-doom#806]
I've had a look at how PrBoom+ achieves demo compatibility with Doom v1.2 and found the changes are rather trivial[*] with two exceptions:
1) We need to copy over the entire `P_CheckSight()` with all dependencies from Heretic.
2) We will have to decide if we want auto-detection of v1.2 demos (which appears rather tricky and only possible on the basis of plausibility checks) or if we want the user to explicitely enable v1.2 support (and then support nothing else in this mode).

[*] I don't mean "trivial to find out" but "trivial to implement once known" here!

**Comment by fabiangreffrath** [comment]
Ah, and we possibly don't want to support recording of v1.2 demos at all.

**Comment by jmtd** [comment]
> In this context, I could try to port the v1.2 compatibility code over from PrBoom+ to Choco. Are we interested in that?

Absolutely! I think this has always been on the roadmap, but not started.

Pitch shifting should be enabled by default if the user has not expressed it to be off explicitly and 1.2 mode is enabled. This probably means we need a three-way option for it, on/default-off/off rather than just on/off.

**Comment by fabiangreffrath** [comment]
> Pitch shifting should be enabled by default if the user has not expressed it to be off explicitly and 1.2 mode is enabled. This probably means we need a three-way option for it, on/default-off/off rather than just on/off.

This is already there. The default value for that variable is `-1` which defaults to `0` in Doom and `1` in Heretic and Hexen. This is already taken care of in my v12 branch.

**Comment by aynrandjuggalo1** [comment]
> In this context, I could try to port the v1.2 compatibility code over from PrBoom+ to Choco. Are we interested in that?

I think this would be awesome. One step closer to full compatibility with Maximum Doom! ;-)

**Comment by chungy** [comment]
> Ah, and we possibly don't want to support recording of v1.2 demos at all.

I don't see why not.

**Comment by AXDOOMER** [comment]
> Ah, and we possibly don't want to support recording of v1.2 demos at all.

Do a full emulation, not just half of it. It's gonna suck if it can only play demos, but not record them. Even bugs that don't impact demo compatibility such as visual glitches should be reproduced.

**Comment by fragglet** [comment]
The nature of demo sync is that if you can play demos back reliably it's probably trivial to do recording too.

**Comment by chungy** [comment]
1.2 emulation probably makes it not a very big stretch to aim for 0.99/1.0 and 1.1 either. It would be pretty nice to have Chocolate Doom capable of playing the original shareware!

**Comment by fabiangreffrath** [comment]
> This is already taken care of in my v12 branch.

https://github.com/fabiangreffrath/crispy-doom/tree/v12

**Comment by fabiangreffrath** [comment]
One problem is that the v1.2 IWAD is missing some lumps that are required by the engine, e.g. 
```
W_GetNumForName: STTMINUS not found!
```

What shall we do about this?

**Comment by haleyjd** [comment]
You can't really separate 1.2 demo support from general 1.2 support IMO.

**Comment by AXDOOMER** [comment]
> What shall we do about this?

Why would a v1.2 emulation need STTMINUS? You could't have a negative score in deathmatch and that's why it wasn't there in this version.

**Comment by fabiangreffrath** [comment]
I invite anyone to work further on my current branch, which I have pushed as
https://github.com/chocolate-doom/chocolate-doom/tree/v12

I don't say that anything is already perfect on this branch. The v1.2 IWAD works if it is loaded as a PWAD alongside an Ultimate DOOM.WAD (because of some missing lumps) and the first demo already gets pretty far before it desyncs. I haven't checked yet how well it works in PrBoom+, though, or how well other demos stay in sync.

In fact, this still requires a lot of work and I am not sure if I have the time and motivation to stem this on my own in the short term,

**Comment by AXDOOMER** [comment]
If v1.2 compatibility ever becomes part of Chocolate-Doom, then we must change some static limits to something lower according to v1.2's limits and Fraggle may not agree with that since he strongly disagrees with raising them to support Doom-plus.

**Comment by CapnClever** [comment]
Complicating the code isn't as high of a priority as the emulation itself. I wouldn't be surprised if **every** codepath involving emulation complicates the code: it's done because one of the most important features of Chocolate Doom (probably the most important from the perspective of comparing source ports) is to emulate vanilla. Raising limits *vis a vis* Doom-plus doesn't emulate vanilla, so it's considered via other factors.

On the other hand, one might be able to more effectively argue for the adjustment of certain limits once there's code in Chocolate Doom in place to do so, which may arise as a result of introducing additional version compatibility. I still wouldn't expect anything (breaking from vanilla gameplay is a heavy factor against it) but it's... less complicated, at least.

**Comment by AXDOOMER** [comment]
> Raising limits vis a vis Doom-plus doesn't emulate vanilla

> breaking from vanilla gameplay is a heavy factor against it

How does adding support for this hack breaks vanilla gameplay? Emulating Doom Plus is just the same as emulating the vanilla engine patched with the longtics hack, which was added Chocolate-Doom, although it breaks vanilla gameplay of course. 

I just feel like **there is a double standard**. 

If Doom Plus "doesn't emulate vanilla", then how does longtics emulate vanilla? It was added to Chocolate-Doom, so that argument doesn't work. 

I don't see the logic in there. The only reason I would see is that it would make the code too much complicated, but I did a source port that does exactly like this hack and you can disable the raised limits without any problem and the code is not that much ugly. I just made the static arrays big enough, added a file with the factors used to multiply the initial limits and added some ternary ifs when the game checks for overflows to know which limit to use if Doom Plus emulation is enable. They don't make Chocolate-Doom slower in practice either. 

Of course, these changes can be used to emulate Doom v1.2 limits. 

https://github.com/AXDOOMER/chocolate-doom-plus

**Comment by chungy** [comment]
If I'm not mistaken, the longtics hack predated Chocolate Doom and was already common long before the inception of this port. It "merely" fixes the demo recording and playback functionality to have the same turning resolution as gameplay without the patch.

Doom-Plus, on the other hand, didn't appear on the scene until *after* Chocolate Doom was already a thing, it has never been a common hack to use (Why would it? It's still in DOS!), and its mainy purpose was that some limit-removing maps would still work in expanded limits. Nobody actually maps with the Doom-Plus limits in mind, but rather the vanilla limits or no limits at all. Crispy Doom well fulfills playing limit-removing maps without breakages just because a map uses a few more visplanes than Doom-Plus supports.

Chocolate Doom expanding to 1.2 emulation, on the other hand, seems totally fair to have it emulate the old limits rather than the 1.9 limits. It's really not the same as expanding out into Doom-Plus' (somewhat arbitrary) limits.

**Comment by AXDOOMER** [comment]
The original longtics hack came out in 2004.
Chocolate-Doom came out in 2005.
The Doom-Plus hack came out in 2006. 

If the Doom-Plus hack would have been released a year before Chocolate-Doom, would it have been already added to the port? What makes the longtics hack a more common hack than Doom-Plus? There are a lot people asking to support Doom-Plus.

I understand the hack raises the limits to arbitrary values, but it's still a hack that was released for vanilla Doom, at the same extent as longtics and novert. It wasn't made just so that people could ask for this feature to be added to Chocolate-Doom some time later.

**Comment by AXDOOMER** [comment]
I think judging a hack by its popularity, its release date or by what it does is missing the point. If it's a hack that was made for vanilla Doom, then it should be integrated into Chocolate-Doom, because that's emulating vanilla, else Chocolate-Doom is not making all that was possible with vanilla possible with itself.

**Comment by CapnClever** [comment]
This really isn't the place to be discussing limit-raising/removal. My main point was that there's abundant reason to adjust limits to fit v1.2, an official version of vanilla, as opposed to a feature that extends vanilla. The side-comment about such changes possibly affecting the decision of including other features was evidently a poor decision, and for that I apologize.

If you want to continue limit-related discussion, please do it in a more relevant thread like #427. There's already more than enough related to v1.2 that merits discussion without going off-topic.

**Comment by fragglet** [comment]
I understand how it may seem a double standard that Chocolate Doom supports `-longtics` but not Doom+'s raised limits. The fact is, raised limits just don't fit with my vision of the project. `-longtics` was a nice, simple feature to add in the early days of the project, but when it became clear that people were doing more elaborate hacks of the vanilla DOS executables I had to draw a line - otherwise, anyone can make a vanilla hack and someone can make an argument that Chocolate Doom should support it. I've always felt that having an intentionally limited scope for this project is important. As I've made clear in the past, there are plenty of spinoff ports that provide this functionality (raised limits) if you really want it.

**Comment by fabiangreffrath** [comment]
Lowering the limits basically means you can leave the static arrays as they are and merely need to adjust the tests that check if the limits have been reached.

This makes it a completely different pair of shoes than actually *increasing* the static array size as support for Doom-Plus limits would require (and which is again lively debated in https://github.com/chocolate-doom/chocolate-doom/issues/427 ).
